[yisiox] iP

yisiox opened the pull request:
# Earl
> An earl is a member of the peerage, ranking below a marquess who ranks below a duke.

Earl keeps track of what you need to remember for you, including
- Todos
- Deadlines
- Events

Earl comes with the functionality to 
- store date and time information
- mark tasks as completed
- save data across sessions
- search for tasks by keyword

## Getting Started
1. Download it from [here](https://github.com/yisiox/ip/releases)
2. Run the jar file
3. Start tracking your tasks

And it is **FREE**!

## Commands
```
list
mark <item no>
unmark <item no>
todo <description>
deadline <description> /by <dd/MM/yyyy> <HHmm>
event <description> /from <dd/MM/yyyy> <HHmm> /to <dd/MM/yyyy> <HHmm>
delete <item no>
find <keyword>
```
### Upcoming Features
- [ ] GUI
- [ ] ~~help command~~ (might not happen 😢 )

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

yisiox committed:
Changed Duke to Earl

yisiox committed:
Implemented Echo and added generic response function

yisiox committed:
Added list functionality

yisiox committed:
Added Task abstraction and ability to mark tasks

yisiox committed:
Added derived classes of Task

yisiox committed:
Added semi-automated testing

yisiox committed:
Added basic input validation

yisiox committed:
Added functionality to delete

yisiox committed:
Modify style to fit style guide

yisiox committed:
Add save functionality

yisiox committed:
Merge save feature

yisiox committed:
Add support for date and time

yisiox committed:
Merge Level-8 into master

yisiox committed:
Refactor to follow more OOP

The Handler class is easily extensible to dispatch a handler for
new behaviour. Role of parsing, storage management and user
interaction is abstracted.

yisiox committed:
Fix to adhere to abstraction barrier

yisiox committed:
Refactor as package earl

yisiox committed:
Merge to add gradle support

yisiox committed:
Move handler into logic package

yisiox committed:
Remove circular dependency

yisiox committed:
Handler: refactor as multiple files

yisiox committed:
Refactor handlers to be more modular

yisiox committed:
Add JUint tests

yisiox committed:
Add JavaDoc to most classes

yisiox committed:
Update minor changes to style

yisiox committed:
Add search for keyword functionality

yisiox committed:
Merge JavaDoc comments

yisiox committed:
Merge coding standard changes

yisiox committed:
Merge search functionality

yisiox committed:
FindHandler: minor change of comments

yisiox committed:
Add factory classes to remove circular dependency

yisiox committed:
Fix unused import statements

yisiox committed:
Add checkstyle and fix style

yisiox committed:
Merge style changes

yisiox committed:
Set up gradle and some files

yisiox committed:
Parser: overload parseUserInput

yisiox committed:
Merge Parser update

yisiox committed:
Implement minimally working GUI

yisiox committed:
Improve GUI functionality

yisiox committed:
Update wider window to prevent wrapping

yisiox committed:
Merge with GUI branch

yisiox committed:
Update GUI font

yisiox committed:
Add font file

jocelyn-soh commented on src/main/java/earl/util/Parser.java:
```diff
@@ -0,0 +1,53 @@
+     * @param input  a line input by the user
+     * @param regex  the regex pattern to split at
+     * @return       an array of {@code String} split as specified
+     */
```
Should the first sentence of the method header comments start in the form of Returns..., Sends..., Adds..., etc.?

yisiox replied:
Thank you for the comment! That is true. I will get to fixing it.

jocelyn-soh commented on src/main/java/earl/gui/DialogBox.java:
```diff
@@ -0,0 +1,65 @@
+import javafx.scene.control.Label;
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.HBox;
```
I like how you imported all of the classes explicitly instead of using wildcards even though you have multiple import statements. The ordering of the import statements are also consistent!

jocelyn-soh commented on src/main/java/earl/tasks/Task.java:
```diff
@@ -0,0 +1,61 @@
+            return true;
+        }
+        return false;
+    }
```
Perhaps the boolean methods can be renamed to sound like booleans, e.g. isMarkedDone, isMarkedUndone.

yisiox replied:
The boolean here refers to the success of changing the completion status of the task; it returns false if it is already of the desired state. I can see that this part may not be clear what it is doing, I will think of a way to improve its clarity. Thanks!

jocelyn-soh commented on src/main/java/earl/util/Storage.java:
```diff
@@ -0,0 +1,77 @@
+     *
+     * @return an {@code ArrayList} of {@code Task} read
+     * @throws EarlException if the file could not be created or read
+     */
```
I like that your JavaDoc comments are well elaborated for the more complex methods!

jocelyn-soh commented on src/main/java/earl/Earl.java:
```diff
@@ -0,0 +1,94 @@
+
+    /**
+     * Entry point of the Earl class. File path is decided here.
+     */
```
Should the first sentence of the method header comments start in the form of Returns..., Sends..., Adds..., etc.? I have noticed the same issue in some other places too.

jocelyn-soh reviewed (commented):
Code generally adheres to coding standard. Other than a few nits, code LGTM. Good job!

logical-1985516 commented on src/main/java/earl/Earl.java:
```diff
@@ -0,0 +1,94 @@
+    public Earl(String filePath) {
+        ui = new Ui();
+        storage = new Storage(filePath);
+        TaskList temp;
```
Should the variable name be changed to a more descriptive name? The use of 'temp' is also seen in a few other places.

yisiox replied:
Thank you for the comments! I'll get to fixing it.

logical-1985516 commented on src/main/java/earl/tasks/Todo.java:
```diff
@@ -0,0 +1,24 @@
+        super(status, description);
+    }
+
+    @Override
+    public String toString() {
+        return "[T]" + super.toString();
+    }
+
+    @Override
+    public String toStorageString() {
+        return "T," + super.getStatusIcon() + "," + description;
+    }
+}
```
Good job on distinguishing between these 2 methods!

logical-1985516 commented on src/main/java/earl/gui/MainWindow.java:
```diff
@@ -0,0 +1,64 @@
+        earl = e;
+    }
+
+    public void init() {
+        dialogContainer.getChildren().addAll(
+                DialogBox.getDukeDialog(earl.getResponse(), dukeImage));
+    }
```
Can the method name be more descriptive?

logical-1985516 reviewed (commented):
Other than a few nits, LGTM. Well done!

yisiox committed:
Add assertions

yisiox committed:
Separate Parser into more specific classes

To allow Parser methods to return a Handler or Task object directly
without circular dependency. A marker interface is used as the Parser
objects do not share any methods.

Add use of enums for Task and Handler

Enum factory design pattern is used for Task and Handler object
creation for added clarity and type safety. Replaces use of switch
cases in Factory objects.

yisiox committed:
Improve code quality

SLAP was followed and happy paths made prominent in order to improve
code readability.

Update unit tests

Unit tests were broken by refactoring of code. New tests have been
written that are now relevant.

yisiox committed:
Merge pull request #1 from yisiox/branch-A-Assertions

Add assertions

yisiox committed:
Resolve merge conflicts

yisiox committed:
Merge with assertions

Most assertions have been removed as code has been restructured to
have no impossible paths. Future possibility of using assertion to
check postconditions.

yisiox committed:
Merge pull request #2 from yisiox/branch-A-CodeQuality

Update code quality

yisiox committed:
Add use of streams

yisiox committed:
Merge pull request #3 from yisiox/branch-A-Streams

Add use of streams

yisiox committed:
Encapsulate indexing within TaskList

yisiox committed:
Fix use of stream to store tasks

yisiox committed:
Made Storage generic

yisiox committed:
Add MassOperable interface to support mass operations

yisiox committed:
Implement mass operation functionality.

Mark, unmark and delete are mass operable. A dedicated interval
parser can be used to interpret arguments of ranges like 1 3-5
for convenience.

yisiox committed:
Merge MassOps

yisiox committed:
Add some JavaDoc and style fix

yisiox committed:
Add github action for A-CI

yisiox committed:
Fix accident on A-CI

yisiox committed:
Fix unit tests again

yisiox committed:
Modify Ui to be responsible to format exceptions

To be consistent across operating systems, Ui now checks for the
operating system's line separator and standardises all tabbing
to 4 spaces.

yisiox committed:
Fix unit tests again

yisiox committed:
Change to pass stylecheck

yisiox committed:
Fix runtest.sh

yisiox committed:
Fix runtest.sh again

yisiox committed:
Limit CI to only test windows

yisiox committed:
Remove IO redirection test

yisiox committed:
Change all parsers to having a static parse

yisiox committed:
Add JavaDoc to functional interface

yisiox committed:
Change StorageParser to an interface

yisiox committed:
Change MassOperable to abstract class

yisiox committed:
Storage: fix bug due to lazy stream

yisiox committed:
MassOperableHandler: minor fix to style

yisiox committed:
Change Ui to better GUI UX

yisiox committed:
MainWindow: change duke to earl

yisiox committed:
Change stylesheets for better GUI

yisiox committed:
Merge better GUI

yisiox committed:
Add noble personality

yisiox committed:
Merge personality update

yisiox committed:
Add fancy font to GUI

yisiox committed:
Fix IO redirection test again

yisiox committed:
Fix todo bug

yisiox committed:
Add user guide for earl

yisiox committed:
Fix Ui.png

yisiox committed:
Improve README.md

yisiox committed:
Fix table in docs

yisiox committed:
Add more unit tests

yisiox committed:
Merge added unit tests

yisiox committed:
Add more error handling

Separated out all Earl style responses from utility classes. Only
Handlers and Earl will generate user text. Added StorageException
and ParserException to replace previous use of EarlException in
such cases.

yisiox committed:
Merge error handling improvement

yisiox committed:
Change to SLAP hard

yisiox committed:
Minor coding standard fix

yisiox committed:
Bugfix for release

yisiox committed:
Update JUnit tests

yisiox committed:
Minor bug fix for release

yisiox committed:
Add check for temporal error

yisiox committed:
Delete unused imports

yisiox committed:
Remove IORedirection tests

yisiox committed:
TaskList: removed dead code

yisiox committed:
Ui: remove Earl specific behaviour

Moved welcome and goodbye message definitions into the Earl class
instead.

Add HelpHandler

A help command is invoked when the user types "help", or an invalid
input, enumerating the list of available commands.

yisiox committed:
README: update new command and formatting

yisiox committed:
Improve readability of mass operations

yisiox committed:
Change font to be more readable

yisiox committed:
Change font to be more readable again

yisiox committed:
Expand test coverage

yisiox committed:
Fix code style

yisiox committed:
Fix formatting in GUI

yisiox committed:
Fix temporal bug

yisiox committed:
Add AI assisted docs
